compiler/src/dotty/tools/dotc/transform/LazyVals.scala

```diff
@@ -274,171 +288,218 @@ class LazyVals extends MiniPhase with IdentityDenotTransformer {
-    *     }
-    *   }
-    * }
-    * ```
-    */
+  /**
+   * Create a threadsafe lazy accessor equivalent to the following code:
+   * ```
+   * private @volatile var _x: AnyRef = null
+   * @tailrec def x: A =
+   *    _x match
+   *    case null =>
+   *        if CAS(_x, null, Evaluating) then
+   *            var result: AnyRef = null // here, we need `AnyRef` to possibly assign `NULL`
+   *            try
+   *                result = rhs
+   *                nullable = null // if the field is nullable; see `CollectNullableFields`
+   *                if result == null then result = NULL // drop if A is non-nullable
+   *            finally
+   *                if !CAS(_x, Evaluating, result) then
+   *                    val lock = _x.asInstanceOf[Waiting]
+   *                    CAS(_x, lock, result)
+   *                    lock.release()
+   *        x
+   *    case Evaluating =>
+   *        CAS(_x, Evaluating, new Waiting)
+   *        x
+   *    case current: Waiting =>
+   *        current.awaitRelease()
+   *        x
+   *    case NULL => null
+   *    case current: A =>
+   *        current
+   * ```
+   * Where `Evaluating` and `NULL` are represented by `object`s and `Waiting` by a class that
+   * allows awaiting the completion of the evaluation. Note that since tail-recursive
+   * functions are transformed *before* lazy-vals, this implementation directly implements
+   * the resulting loop. `PatternMatcher` coming before `LazyVals`, the pattern matching block
+   * is implemented using if-s. That is:
+   * 
+   * ```
+   * private @volatile var _x: AnyRef = null
+   * def x: A =
+   *   while true do
+   *     val current: AnyRef = _x
+   *     if current == null then
+   *       if CAS(_x, null, Evaluating) then
+   *         var result: AnyRef = null
+   *         try
+   *           result = rhs
+   *           nullable = null
+   *           if result == null then result = NULL
+   *         finally
+   *           if !CAS(_x, Evaluating, result) then
+   *             val lock = _x.asInstanceOf[Waiting]
+   *             CAS(_x, lock, result)
+   *             lock.release()
+   *     else
+   *       if current.isInstanceOf[Evaluating] then
+   *         CAS(current, Evaluating, new Waiting)
+   *       else if current.isInstanceOf[NULL] then
+   *         return null
+   *       else if current.isInstanceOf[Waiting] then
+   *         current.asInstanceOf[Waiting].awaitRelease()
+   *       else
+   *         return current.asInstanceOf[A]
+   *   end while
+   * ```
+   * 
+   * @param methodSymbol  the symbol of the new method
+   * @param claz          the class containing this lazy val field
+   * @param target        the target synthetic field
+   * @param rhs           the right-hand side expression of the lazy val
+   * @param tp            the type of the lazy val
+   * @param offset        the offset of the field in the bitmap
+   * @param getFlag       a flag for the volatile get function
```

Comment:
Why all these additional parameters? Can't they be computed in the function themselves?